x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -0,0 +1,151 @@
+
+        // for a minimal restore we basically disable indexing on all fields and only create an index
+        // that is fully functional from an operational perspective. ie. it will have all metadata fields like version/
+        // seqID etc. and an indexed ID field such that we can potentially perform updates on them or delete documents.
+        ImmutableOpenMap<String, MappingMetaData> mappings = snapshotIndexMetaData.getMappings();
+        Iterator<ObjectObjectCursor<String, MappingMetaData>> iterator = mappings.iterator();
+        IndexMetaData.Builder builder = IndexMetaData.builder(snapshotIndexMetaData);
+        while (iterator.hasNext()) {
+            ObjectObjectCursor<String, MappingMetaData> next = iterator.next();
+            MappingMetaData.Routing routing = next.value.routing();
+            final String mapping;
+            if (routing.required()) { // we have to respect the routing to be on the safe side so we pass this one on.
+                mapping = "{ \"" + next.key + "\": { \"enabled\": false, \"_meta\": " + next.value.source().string() + ", " +
+                    "\"_routing\" : { \"required\" : true } } }";
+            } else {
+                mapping = "{ \"" + next.key + "\": { \"enabled\": false, \"_meta\": " + next.value.source().string() + " } }";
+            }
+            builder.putMapping(next.key, mapping);
+        }
+        builder.settings(Settings.builder().put(snapshotIndexMetaData.getSettings())
+            .put(SOURCE_ONLY.getKey(), true)
+            .put("index.blocks.write", true)); // read-only!
+        return builder.build();
+    }
+
+    @Override
+    public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId, IndexId indexId, IndexCommit snapshotIndexCommit,
+                              IndexShardSnapshotStatus snapshotStatus) {
+        if (shard.mapperService().documentMapper() != null // if there is no mapping this is null
+            && shard.mapperService().documentMapper().sourceMapper().isComplete() == false) {
+            throw new IllegalStateException("Can't snapshot _source only on an index that has incomplete source ie. has _source disabled " +
+                "or filters the source");
+        }
+        ShardPath shardPath = shard.shardPath();
+        Path dataPath = shardPath.getDataPath();
+        // TODO should we have a snapshot tmp directory per shard that is maintained by the system?
+        Path snapPath = dataPath.resolve(SNAPSHOT_DIR_NAME);
+        try (FSDirectory directory = new SimpleFSDirectory(snapPath)) {
+            Store tempStore = new Store(store.shardId(), store.indexSettings(), directory, new ShardLock(store.shardId()) {
+                @Override
+                protected void closeInternal() {
+                    // do nothing;
+                }
+            }, Store.OnClose.EMPTY);
+            Supplier<Query> querySupplier = shard.mapperService().hasNested() ? Queries::newNestedFilter : null;
```

Comment:
I think we need to filter soft deletes too - I hope this would be taken care of by the lucene level readers in the SourceOnlySnapshot but wanted to double check.

Reply:
yes it's been taken care of. The soft-deletes field is recorded in the index and we read that.